### ISSUE TITLE
onChange and onSelect shouldn't fire if value is changed programmatically

### ISSUE THREAD

**Issue by dalmo3** [iambumblehead/react-dropdown-now#96]
In a scenario where the dropdown value is a state variable that can also be set by another component, whenever you change the variable `react-dropdown-now` will fire both `onChange` and `onSelect`. 

If the dropdown also sets the very same variable (arguably the main use case for a dropdown), you'll get double updates. Not to mention unwanted side effects.

A very simple case is one where you want to reset the dropdown. See sandbox: https://codesandbox.io/s/thirsty-microservice-kj4sv. I also added a possible workaround by testing if the variable is being changed. The bottom dropdown works as expected.

I believe that _onSelect_ by its very name implies the user executing an action. On the other hand I can see how `onChange` could be more flexible in that front, being allowed to fire by external changes, but the fact I needed to add an equality check to prevent it from running is a bit ironic given its name.

**Comment by dalmo3** [comment]
Also just found out that `onChange` fires when the component is unmounted too!?

**Comment by iambumblehead** [comment]
@dalmo3 there is a pending MR to update some dependencies. When its merged I'll try to resolve the issue you're describing here.

If you're available and would be willing to review my MR, reply here and I will assign the MR to you for review.

**Comment by iambumblehead** [comment]
@dalmo3 the PR is here https://github.com/iambumblehead/react-dropdown-now/pull/102

I did not reproduce the unmounted event and if I missed some detail in your original message, feel free to reply and let me know.

**Comment by iambumblehead** [comment]
@dalmo3 the issues seem to be resolved. Use the onChange event for the discerning onSelect behaviour you need and feel free to re-open this once the changes are published and the ticket closed.

**Comment by iambumblehead** [comment]
version 6.0.0 is published https://www.npmjs.com/package/react-dropdown-now

closing

**Comment by dalmo3** [comment]
That seems like an good solution. Thanks for your work mate!

However now it throws `Invalid hook call`. That happens even in 5.1.1 too which suggests it could be a dependency issue. [Sandobx](https://codesandbox.io/s/recursing-hooks-r6dqg?file=/src/App.js:0-193)

**Comment by iambumblehead** [comment]
@dalmo3 thanks for follow up. I'll respond in the next day or two.

**Comment by iambumblehead** [comment]
@dalmo3 thank you for the great demo. There seems to be a conflux of issues here.

react-dropdown-now should be updated to react 17 but "storybook" used to preview the dropdown component basically depends on react 16 and does not allow react 17 (the storybook maintainer proposes a work-around https://github.com/reach/router/pull/432#issuecomment-734891630)

However, that's only _part_ of the issue. I noticed if I change your example project to use react and react-dom 16 the error is still there. I'm not sure how to proceed atm but am thinking about it.

cc @lwhiteley (and @dalmo3 ) I would be glad to have your advice or a recommendations

**Comment by iambumblehead** [comment]
A branch updating the dropdown to react 17 encountered issues with storybook reported here: https://github.com/storybookjs/storybook/issues/10662#issuecomment-848278108

**Comment by dalmo3** [comment]
> react-dropdown-now should be updated to react 17 but "storybook" used to preview the dropdown component basically depends on react 16 and does not allow react 17 (the storybook maintainer proposes a work-around reach/router#432 (comment))

That's weird. I have another project with React 17 + Storybook 6.2.9 working fine. That project is running yarn v2 and I wasn't the one who set it up, so I'm not sure if there's was an issue with peer dependencies.

Also the #103 [preview](https://iambumblehead-react-dropdown-now-preview-pr-103.surge.sh/?path=/story/docs-introduction--page) seems to work fine for me. 

I tried running that same minimal project from my last comment locally with yarn (both v1 and v2) and got the same behaviour as the sanbox: 5.0.0 works, 5.1.1 and 6.0.0 don't, irrespective of React version. 

In summary, I guess I can't see the link between this error and the React version - as you also observed.

Btw this should probably be a separate issue from https://github.com/iambumblehead/react-dropdown-now/issues/96#issuecomment-847429642.

**Comment by iambumblehead** [comment]
@dalmo3 thanks for looking at it and the thoughtful response. One of the three possible issues suggested by the warning message is mismatching react versions. 

No, changing the react version does not resolve the issue at the dropdown.

I tried change the way the dropdown component to use a default import and export value but that did not resolve the error. The component and the place where the error message occur look correct to me and I'm not yet sure how it will be resolved.

**Comment by dalmo3** [comment]
Yeah, I'm lost here, sorry. Due to the huge number of dependency updates from 5.0.0 to 5.1.1 maybe the logical step would be slowly rolling back each of them to find the culprit. Or even going back to node 12 for build.

**Comment by iambumblehead** [comment]
@dalmo3 that's a good recommendation

**Comment by iambumblehead** [comment]
![localhost_3000_](https://user-images.githubusercontent.com/2058705/119600824-b38f6880-bd9c-11eb-9f34-ec51b2464183.png)

running a local version @dalmo3's sources shows this in the browser

**Comment by dalmo3** [comment]
Ah, missed that. It's whining about `useMemo`? Interesting that #100 did in fact changed that but it was just formatting!?

I don't see where it violates any rules, but to be fair I don't see the point of `useMemo` there, since the `originalOptions` is an array and coming from a prop that means it will always change on render.

**Comment by lwhiteley** [comment]
Yup that's true. Can be removed actually

**Comment by dalmo3** [comment]
I just did a dirty and quick test editing the `.js` directly

```
   // var options = react.exports.useMemo(function () { return prepareOptions(originalOptions); }, [originalOptions]);
    var options = prepareOptions(originalOptions);
```

Still same error shows up but this time the stack tracer doesn't highlight the `useMemo` line. 

![image](https://user-images.githubusercontent.com/33860060/119604160-b93f8980-be42-11eb-9a05-28438e7de350.png)
This is actually the same thing the sandbox example shows.

**Comment by lwhiteley** [comment]
the issue isnt the useMemo. its that there are two react versions being installed somehow, im looking into it

**Comment by iambumblehead** [comment]
@dalmo3 your sandbox works using the just-now-published react-dropdown-now 5.1.2

I'll try to un-publish 6.0.0 and publish a 6.0.1

**Comment by dalmo3** [comment]
Oh yeah, just tested the sandbox from OP and the original issue is fully solved on 6.0.1. 

I have yet to test on the app I was working on to check the unmounting issue I mentioned, but I believe it was a consequence of the original issue and probably solved as well.

Thank you @iambumblehead and @lwhiteley!
